pkg/core/sync.go

```diff
@@ -0,0 +1,258 @@
+		res.State = Downgraded
+	default:
+		return nil, diag.FromError(fmt.Errorf("sync: unhandled case"), diag.INTERNAL)
+	}
+
+	if res.State != NoChange {
+		dd := syncTables(ctx, sta, cur, want, s.ResourceTables)
+		diags = diags.Add(dd)
+	}
+
+	log.Debug().Stringer("provider", provider).Stringer("state", res.State).Uint64("errors", diags.Errors()).Msg("provider sync complete")
+	return res, diags
+}
+
+func Drop(ctx context.Context, storage database.Storage, pm *plugin.Manager, provider registry.Provider) diag.Diagnostics {
+	log.Warn().Stringer("provider", provider).Msg("dropping provider schema")
+	s, diags := GetProviderSchema(ctx, pm, &GetProviderSchemaOptions{Provider: provider})
+	if diags.HasDiags() {
+		return diags
+	}
+
+	sta, err := state.NewClient(ctx, storage.DSN())
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer sta.Close()
+
+	tx, err := sta.ProviderSync(ctx)
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer tx.Rollback(ctx) // nolint:errcheck
+
+	log.Info().Str("provider", provider.Name).Str("version", provider.Version).Msg("dropping provider tables")
+	if diags := dropProvider(ctx, tx, provider, resourceTableNames(s.ResourceTables), nil, nil); diags.HasErrors() {
+		return diags
+	}
+
+	if err := tx.UninstallProvider(ctx, provider); err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return diag.FromError(err, diag.DATABASE)
+	}
+
+	return nil
+}
+
+func syncTables(ctx context.Context, sta *state.Client, cur, want *state.Provider, resourceTables map[string]*schema.Table) diag.Diagnostics {
+	want.Tables = resourceTableNames(resourceTables)
+	want.Signatures = resourceSignatures(resourceTables)
+
+	var (
+		curTables     map[string][]string
+		curSignatures map[string]string
+	)
+
+	if cur != nil && len(cur.Tables) > 0 { // Old provider with known, valid data
+		curTables = cur.Tables
+		curSignatures = cur.Signatures
+	} else {
+		curTables = want.Tables // Fallback to installed provider tables
+	}
+
+	tx, err := sta.ProviderSync(ctx)
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer tx.Rollback(ctx) // nolint:errcheck
+
+	// If a single SQL fails, all following commands also fail with "current transaction is aborted"
+	if diags := dropProvider(ctx, tx, want.Registry(), curTables, curSignatures, want.Signatures); diags.HasErrors() {
+		return diags
+	}
+	if diags := installProvider(ctx, tx, resourceTables); diags.HasErrors() {
+		return diags
+	}
+
+	if err := tx.UninstallProvider(ctx, want.Registry()); err != nil {
+		return diag.FromError(fmt.Errorf("uninstall provider failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.InstallProvider(ctx, want); err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
```

Comment:
why not have an UpdateProvider ?

Reply:
this works if we have no provider installed (no tables dropped + no row deleted) as well as updating a provider (`cur` can be nil, remember)

having an UpdateProvider would need either `INSERT ON CONFLICT UPDATE` (ugly) or an `UPDATE` statement (which wouldn't work in this case). And since this is all done inside a TX there's no gain, just complications?